packages/melos/lib/src/common/persistent_shell.dart

```diff
@@ -0,0 +1,84 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import '../logging.dart';
+import 'platform.dart';
+import 'utils.dart';
+
+class PersistentShell {
+  PersistentShell({
+    required this.logger,
+    this.workingDirectory,
+  });
+
+  final _isWindows = currentPlatform.isWindows;
+  final MelosLogger logger;
+  final String? workingDirectory;
+  late final Process _process;
+  Completer<void>? _commandCompleter;
+
+  Future<void> startShell() async {
+    final executable = _isWindows ? 'cmd.exe' : '/bin/sh';
+
+    _process = await Process.start(
+      executable,
+      _isWindows ? ['/C'] : [],
+      workingDirectory: workingDirectory,
+    );
+
+    _process.stdout.listen((event) {
+      final output = utf8.decode(event, allowMalformed: true);
+      logger.logWithoutNewLine(output);
+      _commandCompleter?.complete();
```

Comment:
It's not necessarily complete here right? This will currently complete after the first line has been send to stdout I think?

Reply:
I did some tests, and without this, some of them failed due to `TimeoutException after 0:00:30.000000: Test timed out after 30 seconds.`.
 
The flow is as follows: when a command is sent using `sendCommand`, the `_commandCompleter` is initialized then the stdout listener completes the `_commandCompleter` when it receives output, indicating that the command has finished executing. A similar process occurs for the stderr listener. 

The wait occurs in the `_awaitCommandCompletion` method.

Reply:
But what I mean is that you might hit this callback multiple times for a command right, if the command prints multiple lines? And in those cases you only want the finish the completer for the last line.

Reply:
Hmm, I see your point now. One way to achieve this is to have some kind of marker to be notified if the command hits the last line. For example:

```dart
  final String _successEndMarker = '__SUCCESS_COMMAND_END__';
  final String _failureEndMarker = '__FAILURE_COMMAND_END__';
  
    void _handleCommandCompletion(String output) {
    if (output.contains(_successEndMarker)) {
      _commandCompleter?.complete();
    } else if (output.contains(_failureEndMarker)) {
      _commandCompleter?.completeError('Command failed');
    }
  }
```

And then, 

```dart
  String _buildFullCommand(String command) {
    final formattedScriptStep =
        targetStyle(command.addStepPrefixEmoji().withoutTrailing('\n'));

    final echoCommand = 'echo "$formattedScriptStep"';
    final echoSuccess = 'echo $_successEndMarker';
    final echoFailure = 'echo $_failureEndMarker';

    if (_isWindows) {
      return '''
        $echoCommand && $command && if %ERRORLEVEL%==0 
            ($echoSuccess) else ($echoFailure || true)
      ''';
    }

    return '''
      eval "$echoCommand && $command && if [ \$? -eq 0 ]; 
      then $echoSuccess; else $echoFailure; fi || true"
    ''';
  }

```


The only problem is that this will be printed out. To circumvent this, we can print the report of all commands only when the `stopShell` method is triggered. Additionally, if the user wants to see all the logs immediately, we could add a `verbose` flag.

Reply:
Isn't the output handled by `logWithoutNewLine`? So you could just skip that one if it detects a line with the markers?

Reply:
@jessicatarra any updates on this? :)

Reply:
Yes! I was able to handle this by using a new method very similar to the `logWithoutNewLine` function. This new method will detect the marker, whether it's a success or failure, and also complete the completer.